features/collection/thresholding/impl/src/main/java/org/opennms/netmgt/threshd/AbstractThresholdEvaluatorState.java

```diff
@@ -173,7 +173,11 @@ public String toString() {
         key = String.format("%d-%s-%s-%s-%s-%s", thresholdingSession.getKey().getNodeId(),
                 thresholdingSession.getKey().getLocation(), threshold.getDsType(),
                 threshold.getDatasourceExpression(), thresholdingSession.getKey().getResource(), threshold.getType());
-
+        // Multiple threshold levels for trigger/rearm may end up with the same key.
+        // Fair to assume that threshold definitions with different threshold levels will have different uei.
+        if(threshold.getTriggeredUEI().isPresent()) {
+            key = String.format("%s-%s", key, threshold.getTriggeredUEI().get());
```

Comment:
Is it likely these `EvaluatorState` objects get constructed frequently? If so we could potentially optimize the key to do only one String allocation instead of two. We could also cache the TTL below.

It may not be worth doing, just pointing it out in case.